Purview Catalog data-plane

yifan-zhou922 opened the pull request:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Please ensure to add changelog with this PR by answering the following questions.
  1. What's the purpose of the update?    
      - [x] new service onboarding 
      - [ ] new API version 
      - [ ] update existing version for new feature 
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify 
  2. When you are targeting to deploy new service/feature to public regions? Please provide date, or month to public if date is not available yet.
  3. When you expect to publish swagger? Please provide date, or month to public if date is not available yet.
  4. If it's an update to existing version,  please select SDKs of specific language and CLIs that require refresh after swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No, no need to refresh for updates in this PR

### Contribution checklist:
- [ ] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [ ] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  -  [ ] Ensure to copy the existing version into new directory structure for first commit (including refactoring) and then push new changes including version updates in separate commits. This is required to review the changes efficiently.
  - Adding a new service

- [ ] Please ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

yifan-zhou922 committed:
update data-plane

jhendrixMSFT commented:
@yifan-zhou922 has this been reviewed by the API review board?

yifan-zhou922 commented:
@jhendrixMSFT Yes, they have reviewed.

iscai-msft committed:
add endpoint to purview swagger, add python config (#13893)

* move readme and add readme.python.md

* add endpoint param

* add endpoint to examples

* fix input file

yifan-zhou922 committed:
update status value

jhendrixMSFT commented:
@yifan-zhou922 can you please fix the spell check CI failures?

jhendrixMSFT commented on specification/purview/data-plane/Azure.Purview.Catalog/preview/2020-12-01-preview/purviewcatalog.json:
```diff
@@ -0,0 +1,6753 @@
+        ],
+        "summary": "Create or update an entity.",
+        "description": "Create or update an entity in Atlas.\nExisting entity is matched using its unique guid if supplied or by its unique attributes eg: qualifiedName.\nMap and array of collections are not well supported. E.g., array<array<int>>, array<map<string, int>>.",
+        "operationId": "EntityREST_CreateOrUpdate",
```
I don't think the suffix "REST" is very helpful.  For context, the operation IDs are used to generate classes and methods.  E.g., in C# this would generate a class (and interface) named `EntityREST` with a method named `CreateOrUpdate`.  Can it be removed?  Comment applies to all operation IDs that contain the "REST" suffix.

yifan-zhou922 replied:
Okay. Will remove the suffix.

yifan-zhou922 committed:
fix spellcheck

yifan-zhou922 committed:
Merge branch 'master' into yifanzhou/purview-catalog-service

yifan-zhou922 commented:
@jhendrixMSFT Spell checks fixed

yifan-zhou922 committed:
remove REST suffix for operationID

yifan-zhou922 committed:
remove REST suffix for operationID

The previous commit includes adding new response which is aligned with the swagger in the repo

yifan-zhou922 committed:
Update purviewcatalog.json

yifan-zhou922 committed:
Update purviewcatalog.json

yifan-zhou922 committed:
Update purviewcatalog.json

yifan-zhou922 committed:
Update purviewcatalog.json

yifan-zhou922 committed:
Revert "Update purviewcatalog.json"

This reverts commit 771caad0a89cca48f188b607ef7fa2f3303d35bd.

yifan-zhou922 committed:
Revert "Update purviewcatalog.json"

This reverts commit e5fae7cd9e0a0fe2ab11c7064ff71308c3726387.

yifan-zhou922 committed:
Revert "Update purviewcatalog.json"

This reverts commit eb33092da92eeabd2e801636bc8bbef88e0cfc3e.

yifan-zhou922 committed:
Revert "Update purviewcatalog.json"

This reverts commit f7a3ba8b78a28168a68b1e426f79dd6292e3cabf.

yifan-zhou922 committed:
Revert "remove REST suffix for operationID"

This reverts commit 3e934eb279202f93d4d4f3158f3069b3e2081c39.

yifan-zhou922 committed:
Update purviewcatalog.json

yifan-zhou922 committed:
remove "json"

yifan-zhou922 committed:
Merge branch 'master' into yifanzhou/purview-catalog-service

yifan-zhou922 committed:
Update purviewcatalog.json

yifan-zhou922 committed:
add suppression OAV108

yifan-zhou922 committed:
update example

yifan-zhou922 committed:
Update readme.md

yifan-zhou922 committed:
Update readme.md

yifan-zhou922 committed:
update suppression D5001

yifan-zhou922 committed:
Update readme.md

jhendrixMSFT commented on specification/purview/data-plane/Azure.Purview.Catalog/preview/2020-12-01-preview/purviewcatalog.json:
```diff
@@ -0,0 +1,6748 @@
+    "/atlas/v2/entity": {
+      "post": {
+        "tags": [
+          "EntityREST"
```
How are these tags used, does the "REST" suffix need to be removed here too?

yifan-zhou922 replied:
Tags are used for grouping operations. We keep "REST" in tags to be compatible with Atlas swagger.

jhendrixMSFT commented on specification/purview/data-plane/Azure.Purview.Catalog/preview/2020-12-01-preview/purviewcatalog.json:
```diff
@@ -0,0 +1,6748 @@
+        ],
+        "summary": "Get relationship information between entities by its GUID.",
+        "description": "Get relationship information between entities by its GUID.",
+        "operationId": "Relationship_GetById2",
```
Why is there a trailing "2"?

yifan-zhou922 replied:
After checking the Atlas swagger, I will move this "2" here.

jhendrixMSFT commented on specification/purview/data-plane/Azure.Purview.Catalog/preview/2020-12-01-preview/purviewcatalog.json:
```diff
@@ -0,0 +1,6748 @@
+            "$ref": "#/parameters/apiVersionAtlas"
+          }
+        ],
+        "responses": {
```
None of the responses define a `default` clause, is that intentional?  The `default` clause can be used to model an error/exception response in the case of failure.  Not mandatory, but if you have a well-defined error response that you want exposed for callers to interact with.

jhendrixMSFT commented on specification/purview/data-plane/Azure.Purview.Catalog/preview/2020-12-01-preview/purviewcatalog.json:
```diff
@@ -0,0 +1,6748 @@
+          }
+        ],
+        "responses": {
+          "200": {
```
Does this actually need to return something, how do callers interact with this result?

yifan-zhou922 replied:
Yes. It will return the deleted entity.

jhendrixMSFT commented on specification/purview/data-plane/Azure.Purview.Catalog/preview/2020-12-01-preview/purviewcatalog.json:
```diff
@@ -0,0 +1,6748 @@
+      "type": "string",
+      "title": "Status",
+      "description": "Status of the AtlasGlossaryTerm",
+      "enum": [
```
All your enum definitions are missing the [x-ms-enum](https://github.com/Azure/autorest/tree/master/docs/extensions#x-ms-enum) attribute.

jhendrixMSFT commented on specification/purview/data-plane/Azure.Purview.Catalog/preview/2020-12-01-preview/purviewcatalog.json:
```diff
@@ -0,0 +1,6748 @@
+        ],
+        "summary": "Create or update entities in bulk.",
+        "description": "Create or update entities in Atlas in bulk.\nExisting entity is matched using its unique guid if supplied or by its unique attributes eg: qualifiedName.\nMap and array of collections are not well supported. E.g., array<array<int>>, array<map<string, int>>.",
+        "operationId": "Entity_CreateOrUpdateBulk",
```
This should probably be `Entity_BulkCreateOrUpdate` to match with the `BulkDelete` operation.

yifan-zhou922 replied:
Rename BulkDelete to Deletebyguids to follow Atlas behavior.

jhendrixMSFT commented on specification/purview/data-plane/Azure.Purview.Catalog/preview/2020-12-01-preview/purviewcatalog.json:
```diff
@@ -0,0 +1,6748 @@
+        ],
+        "summary": "List classifications for a given entity represented by a GUID.",
+        "description": "List classifications for a given entity represented by a GUID.",
+        "operationId": "Entity_GetClassification",
```
Would this be better as something like `EntityClassification_Get`?  Not saying you have to, but at present there are 34 methods on `Entity` which seems a bit fat.  If there's a more logical way to group this it might help make the API surface area easier to reason about.

Comment also applies to `Glossary`

yifan-zhou922 replied:
What we want is to perform get classification operation on entity class, not get operation on classification class. This is designed to follow Atlas behavior.

yifan-zhou922 committed:
Merge branch 'master' into yifanzhou/purview-catalog-service

yifan-zhou922 committed:
update swagger & readme

yifan-zhou922 committed:
remove REST for each example

jhendrixMSFT commented:
I think all that's remaining is cleaning up the enums and fixing the spell check.

yifan-zhou922 commented:
@jhendrixMSFT Hi, will someone check whether csharp & java SDK generation work in readme.md? I realize only the python part has been rewritten.

yifan-zhou922 committed:
update examples

1)  Remove propagate related parameters.
2) update api-version value
3) rename discovery related APIs
4) remove suppression rule R2023

yifan-zhou922 committed:
update swagger

1) update api-version value & name
2) remove tags at the beginning
3) remove REST for each tag
4) remove summary
5) fix grammatical mistake
6) add includeTermHierarchy
7) rename Discovery examples&operationId
8) update several descriptions
9) rename NextLevel-> NextPage
10) rename several parameters
11) add x-ms-enum

jhendrixMSFT commented:
@ShivangiReja @jianghaolu to look at C# and Java READMEs respsectively.

yifan-zhou922 committed:
update api-version in readme

yifan-zhou922 committed:
Update purviewcatalog.json

yifan-zhou922 committed:
Revert "update api-version in readme"

This reverts commit 372b567fc28406aeda53ce861955e6d73cbbe00a.

yifan-zhou922 committed:
update api-version for folder name

yifan-zhou922 committed:
align AutocompleteResult and AutoCompleteRequest

yifan-zhou922 commented:
x-ms-enum added & spell check error fixed.

yifan-zhou922 commented:
@jhendrixMSFT 
Hi, I have a question about the value of "modelAsString" in "x-ms-enum". Do I need to update the value to "false"? For now, the value is "true". But I am not sure whether it is correct. Thanks.

jhendrixMSFT commented:
@yifan-zhou922 for your enums question.  Ideally you should leave it as "true".  This generates the enums in such a way that they're extensible, thus resilient to SDK breaking changes if/when more values are added.

lmazuel committed:
Update custom-words.txt

lmazuel reviewed (approved):
Will have a few naming coming, merging as a baseline

yifan-zhou922 commented:
@lmazuel Hi, this is the PR which renames the namespace to Azure.Analytics.Purview.Catalog https://github.com/Azure/azure-rest-api-specs/pull/14267.
